verifier/src/lib.rs

```diff
@@ -3,10 +3,222 @@
+
+impl<S: Into<String>> From<S> for VerificationError {
+    fn from(message: S) -> Self {
+        Self(message.into())
+    }
+}
+
+/// A verification step. These can chained together using the [`Or`] and [`And`]
+/// types.
+pub trait VerificationStep {
+    /// Performs a verification operation for the [`VerificationStep`].
+    ///
+    /// When verification fails the [`VerificationError`] should contain a
+    /// message communicating the cause of the failure.
+    fn verify(&self) -> Result<()>;
+}
+
+impl Debug for dyn VerificationStep {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VerificationStep").finish()
+    }
+}
+
+/// Will perform a logical and operation for the [`VerificationStep::verify()`]
+/// operation.
+///
+/// This is will be a short circuiting operation. If the `left` side fails
+/// the `right` side will *not* be exercised.
+#[derive(Debug)]
+pub struct And<L, R> {
+    left: L,
+    right: R,
+}
+
+impl<L: VerificationStep, R: VerificationStep> And<L, R> {
+    /// Create a new [`And`] instance
+    ///
+    /// # Arguments:
+    /// * `left` - The left, or first, [`VerificationStep`] to perform. If this
+    ///    fails the `right` will not be attempted.
+    /// * `right` - The right, or second, [`VerificationStep`] to perform.
+    pub fn new(left: L, right: R) -> Self {
+        Self { left, right }
+    }
+
+    /// The left side of this logical and instance.
+    pub fn left(&self) -> &L {
+        &self.left
+    }
+
+    /// The right side of this logical and instance.
+    pub fn right(&self) -> &R {
+        &self.right
+    }
+}
+
+impl<L: VerificationStep, R: VerificationStep> VerificationStep for And<L, R> {
+    fn verify(&self) -> Result<()> {
+        self.left.verify()?;
+        self.right.verify()
+    }
+}
+
+/// Will perform a logical or operation for the [`VerificationStep::verify()`]
+/// operation.
+///
+/// This is will be a short circuiting operation. If the `left` side succeeds
+/// the `right` side will *not* be exercised.
+#[derive(Debug)]
+pub struct Or<L, R> {
+    left: L,
+    right: R,
+}
+
+impl<L: VerificationStep, R: VerificationStep> Or<L, R> {
+    /// Create a new [`Or`] instance
+    ///
+    /// # Arguments:
+    /// * `left` - The left, or first, [`VerificationStep`] to perform. If this
+    ///    succeeds the `right` will not be attempted.
+    /// * `right` - The right, or second, [`VerificationStep`] to perform.
+    pub fn new(left: L, right: R) -> Self {
+        Self { left, right }
+    }
+
+    /// The left side of this logical or instance.
+    pub fn left(&self) -> &L {
+        &self.left
+    }
+
+    /// The right side of this logical or instance.
+    pub fn right(&self) -> &R {
+        &self.right
+    }
+}
+
+impl<L: VerificationStep, R: VerificationStep> VerificationStep for Or<L, R> {
+    fn verify(&self) -> Result<()> {
+        self.left.verify().or_else(|_| self.right.verify())
+    }
+}
```

Comment:
Similar story as `And`, we should be able to build a log message that can say "neither <left> nor <right> was true"

Reply:
With newer logic we should be able to walk the tree and provide display/debug logic.
However the change here seems big enough that I think we follow on with the debug/display stuff